_posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc

```diff
@@ -0,0 +1,391 @@
+[source,java]
+----
+    // Redirect back to a path controlled by the Angular client
+    String redirectPath = "/clientCallback";
+
+    boolean proxy = Boolean.getBoolean("ui.proxy");
+    if (proxy) {
+        redirectPath = "http://localhost:4200" + redirectPath;
+    }
+    resp.sendRedirect(redirectPath);
+----
+
+Also, we need to make the front-end point to `http://localhost:8080/servlet/make-external-call` rather than point to the back-end. To do this we make some changes to
+`DefaultComponent`:
+[source,java]
+----
+@Component({
+  selector: 'app-default',
+  template: `
+    In <b>default</b> component.
+    <a [routerLink]="['/other']">Other</a> |
+    <a [routerLink]="['/rest']">Rest</a> |
+    <a href="{{externalUrl}}">External</a>
+  `,
+  styles: []
+})
+export class DefaultComponent {
+  externalUrl = '/servlet/make-external-call';
+
+  constructor() {
+    if (window.location.port === "4200") {
+      this.externalUrl = "http://localhost:8080" + this.externalUrl;
+    }
+  }
+}
+----
+In the real world I would have used an https://angular.io/guide/build#configuring-application-environments[Angular environment] called something like `proxy` and updated the `script` entry we created in `package.json` to use that. But as there are quite a few files involved in doing that, I have taken a simpler approach to demonstrate the same thing. If the `DefaultComponent` finds it is running on port `4200` it will make the servlet URL point to the back-end server. Otherwise it will attempt to go to the servlet on the Angular CLI server, which of course does not have this.
+
+If you have the Quarkus application from the previous step running we need to stop it so that we can restart it with the system property. Once stopped run:
+[source,bash]
+----
+$mvn package quarkus:dev -Dui.proxy
+----
+
+If you don't have the Angular CLI proxy running from the previous step, run `yarn proxy`.
+
+Now go to http://locahost:4200 and see everything working smoothly.
+
+==== Conclusion
+We have seen how to package an Angular and Quarkus application, and tweaks needed to make it behave in a development environment. The tweaks needed are quite small, and offer great developer productivity when working on your application. You basically just fire up Quarkus with the system property we added (`mvn package quarkus:dev -Dui.proxy=true`), and we do the same for Angular to run it in proxy mode (`yarn proxy`). Now we can just forget about it and modify both back-end and front-end code and see changes happen on the next browser refresh. There is no need to repackage and restart the application with every change done in either place.
+
+Finally, to run this in production you need to run:
+[source,bash]
+----
+$mvn package -Dui
```

Comment:
Add section about native image, this is very important part of Quarkus story.

`mvn package -Dui -Pnative` works fine and you see awesome 
```
Quarkus 0.15.0 started in 0.007s. Listening on: http://[::]:8080
```